publisher/publisher.go

```diff
@@ -251,11 +257,21 @@ func (pub *Impl) singleLogSubmit(
 	ctx context.Context,
 	chain []ct.ASN1Cert,
 	serial string,
-	ctLog *Log) error {
+	ctLog *Log) (*ct.SignedCertificateTimestamp, error) {
 
+	start := time.Now()
 	sct, err := ctLog.client.AddChain(ctx, chain)
+	took := time.Since(start).Seconds()
+	status := "success"
+	defer func() {
```

Comment:
Why defer here? If we only care about the success or failure of the submission function, it would be more straightforward to do one stat increment inside the `if err != nil` clause (error) and one outside of it (success). We could put in a little helper function if we wanted to DRY it up a bit, but I don't think that is strictly necessary.

Deferring the stat increment has a somewhat minor unexpected effect where the success stat doesn't get incremented until after the SA write is successful, which could be a second or two if SA is slow.

Reply:
Mainly just to avoid calling the method twice, I'm fine with just doing that though if people don't really care :/

Reply:
Yeah, I think this is a case where the `defer` seems more "special" and forces you to think a little about what's going on, when we really want something straightforward.

For instance, when I first read this `defer`, I though "Oh, this is intended to let any of the error cases set the status to error, while the main case sets it to success. But in that case, the sense is wrong: The status field should start as error, and be set to success only if none of the error cases apply. Oh, but the second two error cases don't really measure errors during submission, so we wouldn't actually want that." So I'm hoping to short-circuit that thought process for future readers. :-)